### PR TITLE
GH-118943: Fix another race condition when generating jit_stencils.h

### PR DESCRIPTION
**Pull request by hroncok**
Another process might have already moved jit_stencils.h.new

* Issue: gh-118943

### CODE REVIEW

File: Misc/NEWS.d/next/Build/2024-06-18-15-28-25.gh-issue-118943.aie7nn.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Fix a possible race condition affecting parallel builds configured with
+``--enable-experimental-jit``, in which FileNotFoundError could be caused by
```
Comment (Eclips4) [inline Misc/NEWS.d/next/Build/2024-06-18-15-28-25.gh-issue-118943.aie7nn.rst]:
```suggestion
``--enable-experimental-jit``, in which :exc:`FileNotFoundError` could be caused by
```

**Comment by brandtbucher** [comment]
Thanks for this, it seems like a nice fix for the issue.

More generally though, I'm surprised that two `jit_stencils.h` files are being written in parallel like this for the same build. That shouldn't be happening, normally.

This seems to be the source of all of the issues you've been finding. Our Makefile is set up to only run one of these jobs per `make` target... I'm not sure that running several targets (like `regen-all` and `all`) in parallel is really intended. Is that what your build does?

**Review by brandtbucher — APPROVED** [review]

**Comment by hroncok** [comment]
We run regen-all first, then all. Are you interested in full logs?

**Comment by brandtbucher** [comment]
Ah, I think I see the issue! `regen-jit` is part of `regen-all`, but `regen-all` already builds the JIT for other reasons.

I think we should remove `regen-jit` from the `regen-all` target. It's really intended for files that will be checked in and aren't generated as part of the normal build.

**Comment by hroncok** [comment]
This was not merged in time for 3.13.0b3. How can I move this forward?

**Comment by hroncok** [comment]
@brandtbucher Could you please merge this? Is this waiting for something else?

**Comment by hroncok** [comment]
@Yhg1s Hey Thomas. We have carried this patch in Fedora since 3.13.0b2. Is there any chance we could get this merged to at least 3.13.0rc2?

**Comment by Eclips4** [comment]
> @Yhg1s Hey Thomas. We have carried this patch in Fedora since 3.13.0b2. Is there any chance we could get this merged to at least 3.13.0rc2?

I've pinged Brandt in the private core development chat on Discord to get his attention.

**Review by brandtbucher — APPROVED** [review]
Sorry for letting this slip! Looks good to me. One thing, though: can you also remove `regen-jit` from the `regen-all` target in `Makefile.pre.in`?

@Yhg1s I think this is reasonable to backport to the next RC. It's just minor fixes for race conditions in the JIT build (thanks @hroncok for your work to make it available on Fedora).

**Comment by Yhg1s** [comment]
Sure, this is fine to get in 3.13.0rc2.

**Comment by hroncok** [comment]
> One thing, though: can you also remove `regen-jit` from the `regen-all` target in `Makefile.pre.in`?

I'm happy to do that in a separate PR.

**Comment by hroncok** [comment]
See https://github.com/python/cpython/pull/122602

**Comment by hroncok** [comment]
Thanks for the second approval @brandtbucher. Could you please merge this as well?

**Comment by hroncok** [comment]
Thanks.

**Pull request merged**